### ISSUE TITLE
isito-validation initContainer report error when istio-cni is not installed correctly

### ISSUE THREAD

**Issue by cmanzi** [istio/istio#21266]
**Bug description**
When enabling istio-cni-repair on Istio 1.4.5, the istio-validation init container fails. Relevant logs:
```
in new validator: <pod_ip>
Listening on 127.0.0.1:15001
Listening on 127.0.0.1:15006
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
```
A brief examination of the code would indicate to me that the validator never starts a listener on the `IptablesProbePort`, unless I am mistake (https://github.com/istio/istio/blob/1.4.5/tools/istio-iptables/pkg/validation/validator.go#L117 and https://github.com/istio/istio/blob/1.4.5/tools/istio-iptables/pkg/validation/validator.go#L166-L178).

**Expected behavior**
istio-validator validates pod CNI and successfully exits.

**Steps to reproduce the bug**
Install Istio 1.4.5 with CNI and repair enabled. Attempt to launch a pod.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
istio: 1.4.5
kube: 1.15.7

**How was Istio installed?**
helm template

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS - kops

**Comment by towens** [comment]
@rlenglet @lambdai looks like @cmanzi already filed the same.

```
in new validator: 100.100.xx.xx
Listening on 127.0.0.1:15001
Listening on 127.0.0.1:15006
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
```

**Expected behavior**
Pod initializes.

**Steps to reproduce the bug**
This is a test cluster we used the sample certs from the release.

**istioctl version**
```
client version: 1.4.5
control plane version: 1.4.5
data plane version: 1.4.4 (1 proxies), 1.4.5 (2 proxies)
```

**kubectl version**
```
Client Version: v1.15.7
Server Version: v1.15.7
```

```
kubectl create ns istio-system && kubectl create secret generic cacerts -n istio-system \
    --from-file=samples/certs/ca-cert.pem \
    --from-file=samples/certs/ca-key.pem \
    --from-file=samples/certs/root-cert.pem \
    --from-file=samples/certs/cert-chain.pem
```

`istioctl manifest apply -f "${OVERRIDE}"`

**Where the override is the following**
```
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  cni:
    enabled: true
  values:
    cni:
      excludeNamespaces:
       - istio-system
       - kube-system
  unvalidatedValues:
    cni:
      logLevel: info
  coreDNS:
    enabled: true

  gateways:
    components:
      egressGateway:
        enabled: true

  values:
    global:
      proxy:
        accessLogFile: "/dev/stdout"
      # Provides dns resolution for global services
      podDNSSearchNamespaces:
        - global
        - "{{ valueOrDefault .DeploymentMeta.Namespace \"default\" }}.global"

      multiCluster:
        enabled: true

      controlPlaneSecurityEnabled: true

    # Multicluster with gateways requires a root CA
    # Cluster local CAs are bootstrapped with the root CA.
    security:
      selfSigned: false

    gateways:
      istio-ingressgateway:
        autoscaleMax: "20"
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
          service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "http"
          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
      istio-egressgateway:
        env:
          # Needed to route traffic via egress gateway if desired.
          ISTIO_META_REQUESTED_NETWORK_VIEW: "external"

```
This is run after. Multi cluster, replicated control plane. Straight from the Istio docs.
```
# As we are using kubedns v. coredns
apiVersion: v1
kind: ConfigMap
metadata:
  name: kube-dns
  namespace: kube-system
data:
  stubDomains: |
    {"global": ["$(kubectl get svc -n istio-system istiocoredns -o jsonpath={.spec.clusterIP})"]}

```
Successfully returns an ip.

The motivation to upgrade to 1.4.5 was a dream that our traffic latency would be removed. On 1.4.4 post replicating the control plane, we observed significant latency. Not sure if another ticket would be in order. My original comment about it. https://github.com/istio/istio/pull/20130#issuecomment-588100323

Pasted here.

Rolled back to 1.4.4. Significant latency. Multi-cluster. The latency is new, came with replicating the control plane.

```
wrk -t12 -c400 -d10m --latency --timeout 100 --header "Host: a.site.com" http://hash.elb.us-east-1.amazonaws.com:80
  12 threads and 400 connections
  Thread Stats   Avg      Stdev    Max       +/- Stdev
    Latency   146.26ms  383.73ms   10.83s    98.66%
    Req/Sec   293.91       50.60    430.00   76.20%
  Latency Distribution
     50%  103.94ms
     75%  116.23ms
     90%  150.19ms
     99%    1.00s
  2089604 requests in 10.00m, 414.99MB read
Requests/sec:   3482.07
Transfer/sec:    708.13KB
```

Any other info I can supply? Thank you!

**Comment by rlenglet** [comment]
@towens can you confirm that the init container exits after 5 minutes (300 seconds)? That's the timeout hardcoded in the implementation.

@lambdai I suspect that this timeout should be shortened. There is no reason to have a timeout longer than a few seconds. And there is no need to retry connecting (cf. [the infinite loop](https://github.com/istio/istio/blob/1.4.5/tools/istio-iptables/pkg/validation/validator.go#L65-L76)) since we ensure that the server socket is ready to accept connections before we attempt connecting to it.

**Comment by towens** [comment]
@rlenglet Not really :) it looks like it restarts the process.

```
NAME                                  READY   STATUS                  RESTARTS   AGE
pod-744674cb4b-d5wj8   0/2     Init:CrashLoopBackOff   8          56m
```

```
NAME                                  READY   STATUS                  RESTARTS   AGE
pod-744674cb4b-d5wj8   0/2     Init:0/1   9          60m
```

**Comment by rlenglet** [comment]
Yes, when an init container exits, it only results in kubelet restarting that same container, not the pod. So you confirmed that the init container exits after a few minutes every time. The increased delay is due to the backoff.
So the init container seems to work as intended.

**Comment by rlenglet** [comment]
I still think it would be good to reduce the timeout. There shouldn't be a need to retry within the validation code, and 5 minutes is too long. It only delays the time when the repair controller can detect and kill this pod.

**Comment by rlenglet** [comment]
Now the real question is: why didn't the repair daemonset detect that this init container failed validation and kill this pod? 56~60 minutes is too long.
@towens please confirm that the `istio-cni-node` daemonset is running on that node and is ready.
Please confirm that the `istio-cni-node` contains a `repair-cni` container. If that's not the case, you should enable it: https://github.com/istio/installer/blob/release-1.4/istio-cni/templates/daemonset.yaml#L81-L115

**Comment by howardjohn** [comment]
The repair daemonset was not enable by default on 1.4. Its just labeling pods, not killing them

**Comment by howardjohn** [comment]
Well that was my understanding at least... based on the link maybe that is not correct

**Comment by rlenglet** [comment]
@howardjohn ah, you're right. Then the system seems to be working as intended.

**Comment by rlenglet** [comment]
@towens Did you observe this on a newly started node?

**Comment by towens** [comment]
Was pulled away, taking a look now.

**Comment by towens** [comment]
Started rolling the nodes hit an AWS account limit.

The only Istio related label on those pods is presently `security.istio.io/tlsMode=istio`

I'll check the `istio-cni-node` daemonset and update the override to include

```
spec:
  cni:
    enabled: true
  values:
    cni:
      repair:
        enabled: true
      excludeNamespaces:
       - istio-system
       - kube-system
  unvalidatedValues:
    cni:
      logLevel: info
```

Thanks for the links.

**Comment by lambdai** [comment]
The validation works as expected. As a lower priority, cherry picking the configurable timeout to 1.4

@towens Let's see if the repair label works

**Comment by towens** [comment]
Hi,

Here is the updated section of the override.
```
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  cni:
    enabled: true
  values:
    cni:
      repair:
        enabled: true
      excludeNamespaces:
       - istio-system
       - kube-system
  unvalidatedValues:
    cni:
      logLevel: info
```

```
ktl -n istio-system get ds
NAME             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR
istio-cni-node   0         0         0       0            0           beta.kubernetes.io/os=linux
```

This [Install Istio with the Istio CNI plugin](https://istio.io/docs/setup/additional-setup/cni/) doesn't say anything about `kops`  which may not be a problem at all, it is what we're using. The same document says that the CNI is chained.

The networking portion of the template looks like
```
  networking:
    calico:
      crossSubnet: true
      majorVersion: v3
    cni: {}
```
Is it possible the prerequisites for the CNI are not being met?

**Comment by cmanzi** [comment]
To provide some more information:
- This pod that couldn't initialize was running on a node where istio-cni was already up and functioning for other pods (shouldn't the validator have succeeded then?)
- I was running the validator with 'delete pods' set to 'true'
- I tried manually deleting the pod after it was crashing for 5 minutes or so, same effect when it was recreated. 
I've since disabled istio-cni due to that and the issue with statefulsets/mtls not being able to connect to the same pod, and fallen back to the istio-init containers.

**Comment by lambdai** [comment]
@cmanzi istio-cni is not working correctly until a grace period after calico cni is ready. The race condition is not well resolved so the validation is introduced. 
Only istio-cni knows if itself is working correctly. Another strong signal is that other istio-injected-pod on the same kubernetes node was proved working.

You mentioned other pods on that node are ready before you deleted the crashing pod, BTW it is a great move.
1. Is the functioning pod istio-injected?
2. Is there other replicas of the crashing pod? Are they good?
3. Is the functioning pod another replica of the crashing one? If not, are you aware of anything different in terms of istio setting, or pod defined pre-init container?

**Comment by zs-ddl** [comment]
@towens Is there some helpful hints in the `describe` for the DaemonSet?
```
kubectl describe -n istio-system ds
```

**Comment by towens** [comment]
@zs-ddl interesting. Then it loops back to [#14977](https://github.com/istio/istio/issues/14977)

**Comment by tmshort** [comment]
I am running into something similar on k8s 1.15.10, using istio/istio-cni 1.4.7 installed via `helm template`. I am also getting:

`Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect connection refused` in the log multiple times.

- I have deployed istio-cni-node in kube-system, and it is up and running. So, I don't believe this is related to #14977
- The statefulset in question has 3 replicas, but only one is attempting to come up.
- The pod is being labeled with `cni.istio.io/uninitialized=true`, and restarting every ~5 minutes (so repair is on).
- The namespace is using automatic injection.
- The istio-proxy is still initializing, so no logs there.
- Deleting the pod does not allow it to continue.

**Comment by lambdai** [comment]
@tmshort 
Thank you for the comment. The repair and validation is heuristic: they don't understand why istio-cni does not inject the iptables.
The next step is to confirm the iptables
1. if iptables is set up correctly, we should blame istio-validation.
2. if iptables is not set correctly, checkout the logs of ds/istio-cni-node install-cni pod on the node where the failing istio-proxy is running. It's likely istio is not installed with the correct setting with k8s. cniBinDir, cniConfDir...

Any how, you can access the k8s node, run iptables-save in the network namespace of the failing pod 

on the k8s node, run below as root
```
1. find the container id: run `docker ps`
2. figure out the network namespace: `docker inspect --format '{{ .State.Pid }}' {container-id-or-name}`
3. run `nsenter -t {pid} -n sudo iptables-save`
```

Let's start with if the iptables-save container any rule with the name "ISTIO"

**Comment by tmshort** [comment]
Looks as though it was a problem with cniConfDir. I have my logging level at debug. It would be nice if the scripts could indicate a problem such as missing expected configuration files or other abnormalities.

**Comment by nfickas** [comment]
@lambdai I am seeing similar issues while running v1.5.1 on k8s version 1.15.5. I am also getting:

`Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect connection refused`

I am not seeing the pod in question being labelled `cni.istio.io/uninitialized=true`.

**Comment by lambdai** [comment]
@tmshort  Thanks for the update!
> It would be nice if the scripts could indicate a problem such as missing expected configuration files or other abnormalities.
Sorry, not at this moment. The challenge is that the validation pod has very low privilege, which is adequate to validate but not to diagnose the root cause.
Let me see if I can bring more visibility at istio-cni-node daemonset.

**Comment by lambdai** [comment]
@nfickas Can you follow the steps to use admin to verify if iptables is injected correctly? 

See my previous comment https://github.com/istio/istio/issues/21266#issuecomment-611794475

In the last 3 months I haven't confirmed any bug at istio-validation. Istio-validation is delivering the error message but usually the root causes are the cni installation.

**Comment by tmshort** [comment]
> Sorry, not at this moment. The challenge is that the validation pod has very low privilege, which is adequate to validate but not to diagnose the root cause.
Let me see if I can bring more visibility at istio-cni-node daemonset.

That's really where I meant.

**Comment by nfickas** [comment]
@lambdai I just got around to having an administrator do the above iptables-save steps. I am seeing rules in there related to the istio-ingressgateway and other istio services. I am not sure if I am looking for the right things in the iptables dump though. Any suggestions?

**Comment by lambdai** [comment]
@nfickas That's weird. What I am expecting on iptables-save result is something like this. AFAIK there is nothing about particular service in k8s.
My best guess is your mentioned iptables-save is the one under k8s node network space. What we need is the iptables inside the failing pod's network namespace. That's the idea of why we need `docker ps`(assuming docker the is your CRI) and `nsenter`
```
...
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Mon Apr 13 06:31:04 2020
```

**Comment by nfickas** [comment]
Alright, that would make sense. I had the administrator ssh into the node that the failing pod says it is on, in the first step it says find the container id. Am I finding the container id of the istio-validation container?

**Comment by lambdai** [comment]
@nfickas 
`docker ps` 
container_id is the first column
You may need to find your target istio-validation container using your judgement. Usually the pod name is encoded in the output so it's not hard to locate.

**Comment by nfickas** [comment]
@lambdai I think the part that I am not getting is where I am supposed to be doing the docker ps command. I was ssh'd into one of the worker nodes on my cluster and I don't see anything related to istio-validation. I tried exec'ing into the failed pod, but I can't exec into it since it is in a "completed phase". Sorry, that's probably a pretty basic question that I am not getting ahold of here.

**Comment by lambdai** [comment]
> since it is in a "completed phase"
I think that is the reason `docker ps` shows nothing about the container of `istio-validation`. IIRC docker ps shows only running container. Once the container is completed, neither `exec` nor `nsenter` works. 

By default the istio-validation exit immediately if the validation passes, or it will last 5 minutes if the validation keeps failing. The 5 minutes is the window you can access the worker and run `docker ps` and `nsenter`.

BTW: According to my known cases, if you delete the failing pod and when the pod is created again but the validation still fails, 99% of the chances is that istio is not installed with the correct cni setting.
Double check the daemonset log of ds/istio-cni-node

> I am not seeing the pod in question being labelled cni.istio.io/uninitialized=true
This observation also indicates the istio-cni is not running well since you mentioned

**Comment by nfickas** [comment]
I double checked the logs of the daemonset and changed the logLevel to debug: here's what I see 
```
Wrote Istio CNI binaries to /host/opt/cni/bin.
/host/secondary-bin-dir is non-writable, skipping
Using CNI config template from CNI_NETWORK_CONFIG environment variable.
CNI config: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "debug",
  "kubernetes": {
      "kubeconfig": "/etc/cni/multus/net.d/ZZZ-istio-cni-kubeconfig",
      "cni_bin_dir": "/var/lib/cni/bin",
      "exclude_namespaces": [ "istio-system" ]
  }
}
Created CNI config istio-cni.conf
Done configuring CNI.  Sleep=true
WARNING: Configured CNI config file "/host/etc/cni/net.d/istio-cni.conf" preempted by "YYY-istio-cni.conflist".
```

**Comment by lambdai** [comment]
multus as a CNI?
Hmm, I didn't test multus it before. Could it somehow not compatible with the istio?
I see multus reference here: https://istio.io/docs/setup/additional-setup/cni/
@jwendell Could you help @nfickas? Looks like you should have resolve this issue on OpenShift.

**Comment by nfickas** [comment]
So I’m not actually running on OpenShift. I am running on PKS (Pivotal Container Service). Before I upgraded to 1.5.1 I was running the CNI with the settings described in the OpenShift sections for 1.4.3 and that seemed to make it work for that previous version. I had just assumed it would work again for 1.5.1.

**Comment by lambdai** [comment]
@nfickas Please create another issue with explicit PKS and upgrade. With the current title people who are familiar with PKSS is less likely to see it.
As in https://istio.io/docs/setup/additional-setup/cni/ , istio-cni is coupled with k8s provider. 

Some advices since you are on 1.5 (Sorry I didn't get the version before)
1. set repair.deletePods = false and reinstall istio. 
In 1.5 the pod is deleted immediately when istio-validation failed. The above change gives you more time to run docker inspect. By reinstall I mean update the istio-cni related config map and update istio-cni-node daemonset.
2. once 1 is done, try another docker inspect to obtain the iptables inside the pod.
3. You can also run journalctl on that k8s node to find evidence if istio-cni is triggered upon pod creation.

**Comment by lambdai** [comment]
Close this one to avoid discussion exploded on various installation

**Comment by alternaivan** [comment]
> 
> 
> @towens Is there some helpful hints in the `describe` for the DaemonSet?
> 
> ```
> kubectl describe -n istio-system ds
> ```
> 
> For us, we're seeing the following:
> 
> ```
> $ kubectl describe -n istio-system ds
> ...
> Events:
>   Type     Reason        Age                  From                  Message
>   ----     ------        ----                 ----                  -------
>   Warning  FailedCreate  26s (x15 over 108s)  daemonset-controller  Error creating: pods "istio-cni-node-" is forbidden: pods with system-cluster-critical priorityClass is not permitted in istio-system namespace
> $
> $ kubectl version --short
> Client Version: v1.17.0
> Server Version: v1.15.7-gke.23
> ```
> 
> Per the K8s discussion on `system-cluster-critical`, looks like we'll have to wait until GKE supports K8s 1.17 to allow for the priority class to be in another namespace other than kube-system.

Hi,

I've got similar error running on EKS K8s version 1.16 and when upgraded to K8s 1.17 the DaemonSet has started working.

Thanks for the hint! 

Cheers! :)

**Comment by jbhf** [comment]
For what its worth if anyone else comes across this we're using microk8's and I was running into a similar issue where Istio was deployed from a helm template and not the microk8's addon. Adding:

```
values:
   cni:
     repair:
       enabled: true
```

worked, but not at first. Making changes to the Istio operator resource didn't take effect and the istio-cni-node-* daemonset in the kube-system namespace kept showing istio-validator container killing and re-creating (even with repair disabled) causing my app deployment to be stuck in the init state. 

 **Deleting and re-creating the daemonset in kube-system and adding in values.cni.repair.enabled.true from above fixed it.**
